File: web_app/src/api/promotion.service.js
// src/api/promotion.service.js
import axiosClient from '../config/axiosClient';
import { API_ENDPOINTS } from '../config/apiEndpoints';

/**
 * Lấy tất cả danh sách khuyến mãi
 * @returns {Promise} Promise chứa danh sách khuyến mãi
 */
export const getAllPromotions = async () => {
    try {
        const response = await axiosClient.get(API_ENDPOINTS.PROMOTIONS.GET_ALL);
        return {
            success: true,
            data: response.data,
            message: 'Lấy danh sách khuyến mãi thành công'
        };
    } catch (error) {
        console.error('Error fetching promotions:', error);
        throw {
            success: false,
            data: null,
            message: error.response?.data?.message || 'Không thể lấy danh sách khuyến mãi',
            status: error.response?.status
        };
    }
};

/**
 * Tạo khuyến mãi mới
 * @param {Object} promotionData - Dữ liệu khuyến mãi
 * @param {string} promotionData.hotelId - ID khách sạn
 * @param {string} promotionData.code - Mã khuyến mãi
 * @param {string} promotionData.name - Tên khuyến mãi
 * @param {string} promotionData.description - Mô tả khuyến mãi
 * @param {number} promotionData.discountValue - Giá trị giảm giá
 * @param {number} promotionData.minBookingPrice - Giá booking tối thiểu
 * @param {string} promotionData.validFrom - Ngày bắt đầu hiệu lực
 * @param {string} promotionData.validUntil - Ngày kết thúc hiệu lực
 * @param {number} promotionData.usageLimit - Giới hạn sử dụng
 * @param {string} promotionData.promotionType - Loại khuyến mãi
 * @returns {Promise} Promise chứa thông tin khuyến mãi đã tạo
 */
export const createPromotion = async (promotionData) => {
    try {
        // Validate required fields
        const requiredFields = ['hotelId', 'code', 'name', 'discountValue', 'validFrom', 'validUntil'];
        for (const field of requiredFields) {
            if (!promotionData[field]) {
                throw new Error(`Trường ${field} là bắt buộc`);
            }
        }

        const response = await axiosClient.post(API_ENDPOINTS.PROMOTIONS.CREATE, promotionData);
        return {
            success: true,
            data: response.data,
            message: 'Tạo khuyến mãi thành công'
        };
    } catch (error) {
        console.error('Error creating promotion:', error);
        throw {
            success: false,
            data: null,
            message: error.response?.data?.message || error.message || 'Không thể tạo khuyến mãi',
            status: error.response?.status
        };
    }
};

/**
 * Cập nhật khuyến mãi
 * @param {string} promotionId - ID của khuyến mãi cần cập nhật
 * @param {Object} promotionData - Dữ liệu khuyến mãi cần cập nhật
 * @returns {Promise} Promise chứa thông tin khuyến mãi đã cập nhật
 */
export const updatePromotion = async (promotionId, promotionData) => {
    try {
        if (!promotionId) {
            throw new Error('ID khuyến mãi là bắt buộc');
        }

        const response = await axiosClient.put(
            API_ENDPOINTS.PROMOTIONS.UPDATE(promotionId), 
            promotionData
        );
        
        return {
            success: true,
            data: response.data,
            message: 'Cập nhật khuyến mãi thành công'
        };
    } catch (error) {
        console.error('Error updating promotion:', error);
        throw {
            success: false,
            data: null,
            message: error.response?.data?.message || 'Không thể cập nhật khuyến mãi',
            status: error.response?.status
        };
    }
};

/**
 * Lấy thông tin chi tiết khuyến mãi theo ID
 * @param {string} promotionId - ID của khuyến mãi
 * @returns {Promise} Promise chứa thông tin chi tiết khuyến mãi
 */
export const getPromotionById = async (promotionId) => {
    try {
        if (!promotionId) {
            throw new Error('ID khuyến mãi là bắt buộc');
        }

        const response = await axiosClient.get(
            API_ENDPOINTS.PROMOTIONS.GET_BY_ID(promotionId)
        );
        
        return {
            success: true,
            data: response.data,
            message: 'Lấy thông tin khuyến mãi thành công'
        };
    } catch (error) {
        console.error('Error fetching promotion:', error);
        throw {
            success: false,
            data: null,
            message: error.response?.data?.message || 'Không thể lấy thông tin khuyến mãi',
            status: error.response?.status
        };
    }
};

// Export default object chứa tất cả các functions (optional)
const promotionService = {
    getAllPromotions,
    createPromotion,
    updatePromotion,
    getPromotionById
};

export default promotionService;

File: mobile/lib/services/promotion_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'api_config.dart';
import 'token_service.dart';

class PromotionService {
  // Singleton pattern
  static final PromotionService _instance = PromotionService._internal();
  factory PromotionService() => _instance;
  PromotionService._internal();

  // Headers mặc định
  Map<String, String> get _headers => {
    'Content-Type': 'application/json',
    'Accept': 'application/json',
  };

  // Headers có token
  Map<String, String> _headersWithToken(String token) => {
    'Content-Type': 'application/json',
    'Accept': 'application/json',
    'Authorization': 'Bearer $token',
  };

  // ============================================
  // PUBLIC METHODS (không cần authentication)
  // ============================================

  /// Lấy tất cả các chương trình khuyến mãi
  /// GET /api/v1/promotions
  Future<Map<String, dynamic>> getAllPromotions() async {
    try {
      final url = Uri.parse('${ApiConfig.baseUrl}/promotions');
      final response = await http.get(url, headers: _headers);
      final data = json.decode(response.body);
      
      if (response.statusCode == 200 && data['status'] == 'success') {
        return {'success': true, 'data': data['data']};
      } else {
        return {'success': false, 'message': data['message'] ?? 'Unknown error'};
      }
    } catch (e) {
      return {'success': false, 'message': 'Lỗi kết nối: $e'};
    }
  }

  /// Lấy khuyến mãi theo hotel ID
  /// GET /api/v1/promotions/hotel/:hotelId
  Future<Map<String, dynamic>> getPromotionsByHotelId(String hotelId) async {
    try {
      final url = Uri.parse('${ApiConfig.baseUrl}/promotions/hotel/$hotelId');
      final response = await http.get(url, headers: _headers);
      final data = json.decode(response.body);
      
      if (response.statusCode == 200 && data['status'] == 'success') {
        return {'success': true, 'data': data['data']};
      } else {
        return {'success': false, 'message': data['message'] ?? 'Unknown error'};
      }
    } catch (e) {
      return {'success': false, 'message': 'Lỗi kết nối: $e'};
    }
  }

  /// Lấy khuyến mãi theo code
  /// GET /api/v1/promotions/code/:code
  Future<Map<String, dynamic>> getPromotionByCode(String code) async {
    try {
      final url = Uri.parse('${ApiConfig.baseUrl}/promotions/code/$code');
      final response = await http.get(url, headers: _headers);
      final data = json.decode(response.body);
      
      if (response.statusCode == 200 && data['status'] == 'success') {
        return {'success': true, 'data': data['data']};
      } else {
        return {'success': false, 'message': data['message'] ?? 'Unknown error'};
      }
    } catch (e) {
      return {'success': false, 'message': 'Lỗi kết nối: $e'};
    }
  }

  /// Lấy chi tiết một chương trình khuyến mãi
  /// GET /api/v1/promotions/:promotionId
  Future<Map<String, dynamic>> getPromotionById(String promotionId) async {
    try {
      final url = Uri.parse('${ApiConfig.baseUrl}/promotions/$promotionId');
      final response = await http.get(url, headers: _headers);
      final data = json.decode(response.body);
      
      if (response.statusCode == 200 && data['status'] == 'success') {
        return {'success': true, 'data': data['data']};
      } else {
        return {'success': false, 'message': data['message'] ?? 'Unknown error'};
      }
    } catch (e) {
      return {'success': false, 'message': 'Lỗi kết nối: $e'};
    }
  }

  // ============================================
  // AUTHENTICATED METHODS (cần authentication)
  // ============================================

  /// Xác thực mã khuyến mãi
  /// POST /api/v1/promotions/validate
  Future<Map<String, dynamic>> validatePromotion({
    required String code,
    required double bookingTotal,
    String? token,
  }) async {
    try {
      final url = Uri.parse('${ApiConfig.baseUrl}/promotions/validate');
      
      String? authToken = token;
      authToken ??= await TokenService.getToken();
      if (authToken == null || authToken.isEmpty) {
        return {'success': false, 'message': 'Bạn chưa đăng nhập'};
      }

      final response = await http.post(
        url,
        headers: _headersWithToken(authToken),
        body: json.encode({
          'code': code,
          'bookingTotal': bookingTotal,
        }),
      );
      
      final data = json.decode(response.body);
      
      if (response.statusCode == 200 && data['success'] == true) {
        return {'success': true, 'data': data['data']};
      } else {
        return {'success': false, 'message': data['message'] ?? 'Unknown error'};
      }
    } catch (e) {
      return {'success': false, 'message': 'Lỗi kết nối: $e'};
    }
  }

  /// Tạo chương trình khuyến mãi mới (Admin/Hotel Owner)
  /// POST /api/v1/promotions
  Future<Map<String, dynamic>> createPromotion({
    required Map<String, dynamic> promotionData,
    String? token,
  }) async {
    try {
      final url = Uri.parse('${ApiConfig.baseUrl}/promotions');
      
      String? authToken = token;
      authToken ??= await TokenService.getToken();
      if (authToken == null || authToken.isEmpty) {
        return {'success': false, 'message': 'Bạn chưa đăng nhập'};
      }

      final response = await http.post(
        url,
        headers: _headersWithToken(authToken),
        body: json.encode(promotionData),
      );
      
      final data = json.decode(response.body);
      
      if (response.statusCode == 201 && data['success'] == true) {
        return {'success': true, 'data': data['data']};
      } else {
        return {'success': false, 'message': data['message'] ?? 'Unknown error'};
      }
    } catch (e) {
      return {'success': false, 'message': 'Lỗi kết nối: $e'};
    }
  }

  /// Cập nhật chương trình khuyến mãi (Admin/Hotel Owner)
  /// PUT /api/v1/promotions/:promotionId
  Future<Map<String, dynamic>> updatePromotion({
    required String promotionId,
    required Map<String, dynamic> promotionData,
    String? token,
  }) async {
    try {
      final url = Uri.parse('${ApiConfig.baseUrl}/promotions/$promotionId');
      
      String? authToken = token;
      authToken ??= await TokenService.getToken();
      if (authToken == null || authToken.isEmpty) {
        return {'success': false, 'message': 'Bạn chưa đăng nhập'};
      }

      final response = await http.put(
        url,
        headers: _headersWithToken(authToken),
        body: json.encode(promotionData),
      );
      
      final data = json.decode(response.body);
      
      if (response.statusCode == 200 && data['success'] == true) {
        return {'success': true, 'data': data['data']};
      } else {
        return {'success': false, 'message': data['message'] ?? 'Unknown error'};
      }
    } catch (e) {
      return {'success': false, 'message': 'Lỗi kết nối: $e'};
    }
  }

  /// Xóa chương trình khuyến mãi (Admin/Hotel Owner)
  /// DELETE /api/v1/promotions/:promotionId
  Future<Map<String, dynamic>> deletePromotion({
    required String promotionId,
    String? token,
  }) async {
    try {
      final url = Uri.parse('${ApiConfig.baseUrl}/promotions/$promotionId');
      
      String? authToken = token;
      authToken ??= await TokenService.getToken();
      if (authToken == null || authToken.isEmpty) {
        return {'success': false, 'message': 'Bạn chưa đăng nhập'};
      }

      final response = await http.delete(
        url,
        headers: _headersWithToken(authToken),
      );
      
      final data = json.decode(response.body);
      
      if (response.statusCode == 200 && data['success'] == true) {
        return {'success': true, 'message': data['message'] ?? 'Xóa thành công'};
      } else {
        return {'success': false, 'message': data['message'] ?? 'Unknown error'};
      }
    } catch (e) {
      return {'success': false, 'message': 'Lỗi kết nối: $e'};
    }
  }

  /// Lọc khuyến mãi theo điều kiện (Admin/Hotel Owner)
  /// GET /api/v1/promotions/filter
  Future<Map<String, dynamic>> filterPromotions({
    String? status,
    String? code,
    String? startDate,
    String? endDate,
    String? hotelId,
    String? token,
  }) async {
    try {
      final queryParams = <String, String>{};
      if (status != null) queryParams['status'] = status;
      if (code != null) queryParams['code'] = code;
      if (startDate != null) queryParams['startDate'] = startDate;
      if (endDate != null) queryParams['endDate'] = endDate;
      if (hotelId != null) queryParams['hotelId'] = hotelId;

      final url = Uri.parse('${ApiConfig.baseUrl}/promotions/filter')
          .replace(queryParameters: queryParams);
      
      String? authToken = token;
      authToken ??= await TokenService.getToken();
      if (authToken == null || authToken.isEmpty) {
        return {'success': false, 'message': 'Bạn chưa đăng nhập'};
      }

      final response = await http.get(
        url,
        headers: _headersWithToken(authToken),
      );
      
      final data = json.decode(response.body);
      
      if (response.statusCode == 200 && data['success'] == true) {
        return {'success': true, 'data': data['data']};
      } else {
        return {'success': false, 'message': data['message'] ?? 'Unknown error'};
      }
    } catch (e) {
      return {'success': false, 'message': 'Lỗi kết nối: $e'};
    }
  }
}
